### ISSUE TITLE
SA1313: Parameter '__' must begin with lower-case letter.

### ISSUE THREAD

**Issue by yaakov-h** [DotNetAnalyzers/StyleCopAnalyzers#1606]
The following code snippet is OK in StyleCop 4.6, but not in StyleCopAnalyzers:

```
Action<object, object> b = (_, __) => { };
```

The warning output is `warning SA1313: Parameter '__' must begin with lower-case letter`.

StyleCopAnalyzers seems to be OK with a single underscore, but not two underscores. I assume this is due to https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1313ParameterNamesMustBeginWithLowerCaseLetter.cs#L86 explicitly checking for a single underscore.

**Comment by sharwell** [comment]
This is related to (but not a duplicate of) #1343.

**Comment by Maxwe11** [comment]
:+1:

**Comment by nbarbettini** [comment]
Should we allow any multiple of underscore characters to be used? After 2 or 3, it's hard to tell them apart. :grimacing:

**Comment by yaakov-h** [comment]
I'd suggest they ought to be sequential (i.e. `___` can't come after `_` unless `__` is between them), and only allowed if they're unused, perhaps as separate rules. If you have a lambda with more than 3 or 4 unused arguments, I'd say you have bigger problems than just style.

**Comment by GregReddick** [comment]
Ick. I wouldn't say that variables with just the underscore as a name is a good idea. But if the goal is to be 100% compatible with StyleCop classic...fine. I don't think we need rules defining how to use them when they shouldn't be used at all, though.

**Comment by nbarbettini** [comment]
@GregReddick The underscores aren't completely horrible when used as intended - in other words, not used at all. I agree, we shouldn't say how to use them - just let people use `_` (or `__`) to mean "unused variable".

**Comment by yaakov-h** [comment]
Alternatively, create a new rule that variable names cannot be underscores. That way developers can choose whether or not [unused variable] names can be underscores, as a completely independent decision to variable names starting with a lowercase letter.

**Comment by nbarbettini** [comment]
Actually, I like that a lot - nice separation of concerns for these rules.

**Comment by sharwell** [comment]
I'm thinking we can defer this topic until after the 1.0 Stable release.

**Comment by yaakov-h** [comment]
Not if 1.0 promises full rule parity with StyleCop - I assume therefore it doesn't?

**Comment by sharwell** [comment]
> Not if 1.0 promises full rule parity with StyleCop - I assume therefore it doesn't?

The primary goals for 1.0 are:

1. *Good* (but not necessarily perfect) compatibility with StyleCop classic.
  * Every StyleCop Classic rule needs either an implementation in StyleCop Analyzers, or a decision made to intentionally not implement the rule.
  * *Documented* behavior must be explicitly addressed, either through implementing the letter of the rules or correcting oversights.
2. Clear documentation of behavior which differs from StyleCop classic.
  * This documentation only needs to document changes in behavior when StyleCop Analyzers on C# 5 (or earlier) code. StyleCop Classic did not support C# 6, so we are defining the behavior on such code for the first time.
3. Reliable code fix behavior, to the maximum extent possible.
  * Code fixes which rename files are not guaranteed to work in shared projects or when files are linked into the project from an external location (multiple external bugs limit our possibilities here)
  * Code fixes which rename *anything* may result in breaking changes, although Roslyn works exceptionally hard to ensure this does not happen
  * All other code fixes cannot result in any breaking changes or loss of data (including trivia, except where that is specifically part of the fix)

The behavior described in this issue is an undocumented feature of StyleCop Classic, which means we need to define it and document it before declaring the situation resolved. This was pretty easy to address for #1343, but as the number of `_` characters grows so does the reluctance of users to make an exception.

**Comment by GraemeBradbury** [comment]
To add my two cents to this.

WPF has a very common type `RoutedEventHandler` so seeing `(_,__) => {...}` when responding to events is rather normal.

While it would be better if C# supported `_` as a syntax for ignored parameters (like other languages). Using multiple underscore to communicate parameters as being unused is still better than using comments to communicate unuse. It also falls into line with alot of the teaching out there for functional programming.

**Comment by ivanlomakov** [comment]
Hi all. So what is the plan for the issue? Is it going to be implemented or not? I guess two underscores need to be allowed at least. I have never met three unused parameters in expression but I have two.

**Comment by vweijsters** [comment]
After reading through the discussion on this issue again, I'd say that allowing two underscores would be good idea (mainly based on the `RoutedEventHandler` argument, but the same would hold true for `EventHandler`). Anything more than that will impact readability and should in my opinion not be allowed.
@sharwell What is your opinion on this?

**Comment by sharwell** [comment]
@vweijsters Sure. And if discards aren't supported by then, we can consider allowing a third underscore in June 2021.

**Comment by jnm2** [comment]
@sharwell PR is up: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/2758

**Comment by chaddrellunlimitied** [comment]
I'm getting this again

SA1313	Parameter '__' must begin with lower-case letter

in 1.0.2

**Comment by chaddrellunlimitied** [comment]
code here,

```
return Policy
                .HandleResult<HttpResponseMessage>(r => 
                        r.StatusCode == HttpStatusCode.Unauthorized ||
                        r.StatusCode == HttpStatusCode.Forbidden)
                .RetryAsync(
                    1,
                    (_, __) => HandleUnauthorized(provider, request));
```

**Comment by yaakov-h** [comment]
The PR to fix was only released with 1.1.0-beta009.

Try using the latest 1.1.118 build.

**Comment by chaddrellunlimitied** [comment]
The stable 1.1.118 fixes this, and is making my code better in new ways.

**Comment by djfr** [comment]
I don't understand the rationale of allowing only up to 2 `_`, but not more. I believe the analyser should validate if the discards in lambdas are incrementing the number of `_` by 1 as they are declared (including chained lambdas), but not stop on 2.

I think this is more readable - assuming the reader knows the `_` discard convention
```c#
(_, __, ___, ____) => { }
```

Then
```c#
(_, __, discard1, discard2) => { }
```

Or even then
```c#
(discard1, discard2, discard3, discard4) => { }
```

**Comment by sharwell** [comment]
@djfr The language team has talked about allowing this syntax, which is what we are hoping to use:

```csharp
(_, _, _, _) => { }
```

Until that's supported, another option is this syntax for cases where all parameters are discards:

```csharp
delegate { }
```
